### PR TITLE
Dynamic 2021

### PR DESCRIPTION
**Pull request by Mothblocks**
## About The Pull Request

Dynamic 2021, among much other things, **splits the threat level into two separate budgets, one for round start and one for midround/latejoin**. You can read the design doc, plus methodology and charts here: https://hackmd.io/@tgstation/S1C4dYJkO.

To server owners: this is **incompatible with current dynamic.json** configurations. This is not just because some fields have different values now, but because the older values are not balanced towards this system.

Inidividual list of changes (as I remember them):
- Threat level is now split into two separate budgets, one for round start and one for midround/latejoin.
- In TESTING, you can now run dynamic simulations to see the roundstart picks.
- `antag_cap` is now based on an equation rather than a fixed list of antags per indice--the old system is not scalable. You can now pass it either a number for a fixed constant, or a list with "denominator" and an optional "offset" to fit the equation `ceil(x / denominator) + offset`, with `x` being population.
- Midround/latejoin timers are now vars on `/datum/game_mode/dynamic` itself, and thus can be configured by head admins per server.
- Removes some older, as far as I can tell completely unused admin tools relating to dynamic.
- `autotraitor_cooldown` is now properly respected from config, and is based on an exact time rather than on ticks (450 used to be 15 minutes, now it's just `15 MINUTES`).
- The rulesets dynamic chooses are now in a `dynamic.json` file in round logs. Relies on https://github.com/tgstation/tgstation13.org/pull/64 to be public.
- Adds more logging to the autotraitor ruleset for admins.
- Midround/latejoin chance modifiers are now configurable.

Labelled "Do Not Merge", as the design doc outlines the specific playtesting schedule I would like to perform.

- [ ] Note to self: Remove the custom MOTD before merge.

## Why It's Good For The Game
(Copied from design doc)

Dynamic currently dumps as much threat as it can into round start threats. This means that rounds at ~50 threat can have an absurd amount of antagonists round start, while leaving very little threat for midround/latejoin antagonists. What inevitably happens is just one of those picked decides to murderbone, leading to the inevitable 20 minute shuttle call. Furthermore, once those antagonists are dead, Dynamic doesn’t have any threat left to spend on new antagonists that keep the round flowing, promoting the mentality of “antags dead, I’m bored, let’s leave”.

The proposed solution of Dynamic 2021 is to split threat level into two separate budgets, one for round start and one for midround/latejoin. This split, like threat level, will also be done on a lorentz curve–meaning that while most rounds will have them evenly split, there is still the possibility of chaotic high antagonist round start rounds, or the possibility of a midround onslaught.

## Changelog
:cl:
add: Dynamic now rolls a separate budget for round start and midround antagonists, meaning less round start antags will roll in favor of more midround ones.
admin: The antag_cap value now uses an equation that scales with population, rather than a fixed array.
admin: autotraitor_cooldown now uses deciseconds rather than ticks, so the older value of 450 now means 45 seconds. Set it to be 9000 if you want it to be accurate to the old value (15 minutes), or simply unset it.
admin: Midround/latejoin dynamic timers are now configurable.
admin: Syndicate Sleeper Agent will now give details when ti fails.
fix: The autotraitor cooldown now respects configuration, rather than resetting back to 15 minutes after it is rolled.
/:cl:

### CODE REVIEW

File: code/game/gamemodes/dynamic/dynamic.dm
Comment (Jared-Fogle) [inline code/game/gamemodes/dynamic/dynamic.dm:318]: This is a TM only change, as per the note under "Playtesting Schedule". It will be reverted when either Dynamic 2021 shows it's no longer necessary to untestmerge completely, or before it is merged.

File: code/game/gamemodes/dynamic/dynamic_simulations.dm
Comment (Jared-Fogle) [inline code/game/gamemodes/dynamic/dynamic_simulations.dm:76]: This wasn't implemented, and will most certainly be removed.

File: code/game/gamemodes/dynamic/dynamic_simulations.dm
Comment (Jared-Fogle) [inline code/game/gamemodes/dynamic/dynamic_simulations.dm:91]: Note to self: I no longer need to do this, the run simulations verb asks for the values,

**Comment by Tlaltecuhtli** [comment]
does it change current odds other than scaling by  pop? mid round antags are usually more boring than round start ones as you might be doing some funny project, get antag and just ignore it cause ur having fun on your own, i dont see any reason to have them more common in exhcange of rounstart one being more rare

**Comment by BadSS13Player** [comment]
Midround antagonists like Syndicate infiltrators or Syndicate sleeping agents are pretty shitty and don't add much to the round, as opposed to midround antags like Xenos, Blob and such. Maybe with an increased midround thread budget it should favor the latter type of antag?

**Comment by Tlaltecuhtli** [comment]
yeah rarer midgame traitor - heretic and more solo midround antags, as late join syndicate just means get out of shuttle and go in escape shuttle cause it already got called and blow up the syndibomb in it

**Comment by Jared-Fogle** [comment]
> does it change current odds other than scaling by pop? mid round antags are usually more boring than round start ones as you might be doing some funny project, get antag and just ignore it cause ur having fun on your own, i dont see any reason to have them more common in exhcange of rounstart one being more rare

> Midround antagonists like Syndicate infiltrators or Syndicate sleeping agents are pretty shitty and don't add much to the round, as opposed to midround antags like Xenos, Blob and such. Maybe with an increased midround thread budget it should favor the latter type of antag?

The midround antags ruining projects is a callout that I've heard before, and on the design doc I outline that I'm sort of on the fence (leaning towards negativity) towards sleeper agent specifically. During playtesting I'm going to look and see if sleeper agent really makes the game better. "Midround antags" in this case means more than just sleeper agent, though. It means the ghost roles (ninja, benos, etc) as well as latejoin antags (such as provocateur or syndicate infiltrator) that don't have these similar quirks.

**Comment by RaveRadbury** [comment]
Is the definition for "midround" going to be a config/could it be?

**Comment by Jared-Fogle** [comment]
@RaveRadbury Can you explain what you mean by that? "Midround" just means anything thats not a roundstart antag, so think blob and latejoin infiltrators.

**Comment by RaveRadbury** [comment]
> @RaveRadbury Can you explain what you mean by that? "Midround" just means anything thats not a roundstart antag, so think blob and latejoin infiltrators.

Apologies! I can see now that was very vague. I was not referring to the antags that fall under the category of midround, rather I was referring to the time that we deem midround to be (20 minutes in, 30 minutes in, etc.)

**Comment by Jared-Fogle** [comment]
Oh, I see. The way it works is after a set period of time in a range (midrounds is 15 to 35 minutes), it'll roll a chance to summon a midround. Those timers used to not be configurable, but now they are.

**Comment by Jared-Fogle** [comment]
Labelling as Do not TM, only for the purposes of that I am going to be at the lead of that (once I get maintainer sign off).

**Comment by optimumtact** [comment]
This needs a readme document explaining how the injection engine works and what the algorithm used for threat spending and injection is, otherwise it's going to bitrot and be hard and confusing to understand and likely replaced at a later date.

And it needs to be good doc too, not some half hearted design overview

Dynamic config is already confusing enough let alone with this on top of it.

**Comment by Jared-Fogle** [comment]
> This needs a readme document explaining how the injection engine works and what the algorithm used for threat spending and injection is, otherwise it's going to bitrot and be hard and confusing to understand and likely replaced at a later date.

I can do this :+1:

**Comment by Jared-Fogle** [comment]
@optimumtact Latest commit updates the current README. Let me know if there is any concerns you have.

File: code/game/gamemodes/dynamic/dynamic_rulesets_midround.dm
Comment (kriskog) [inline code/game/gamemodes/dynamic/dynamic_rulesets_midround.dm:183]: antags -> traitors

Though, do we need these admin messages? Will any admin benefit from constant updates about failures like this? I say we log only. If anything make the log more available via the gamemode panel as an alternative.

File: code/game/gamemodes/dynamic/dynamic_rulesets_midround.dm
Comment (kriskog) [inline code/game/gamemodes/dynamic/dynamic_rulesets_midround.dm:187]: Same as above.

File: code/game/gamemodes/dynamic/dynamic.dm
Comment (kriskog) [inline code/game/gamemodes/dynamic/dynamic.dm:311]: These do not need to be admin messages, its just spammy and is available in game panel + logs if anyone needs to look it up. It cant be changed at this point anyway. Its been on my list forever.

File: code/game/gamemodes/dynamic/dynamic.dm
Comment (kriskog) [inline code/game/gamemodes/dynamic/dynamic.dm:378]: threat_log

**Review by kriskog — COMMENTED** [review]
Some small stuff, otherwise looks fine. Highlander thing isnt a requirement by any means, do it if you feel like it.

File: code/game/gamemodes/dynamic/dynamic_rulesets_midround.dm
Comment (Jared-Fogle) [reply code/game/gamemodes/dynamic/dynamic_rulesets_midround.dm:183]: There's no point to it, no. I'll just move all autotraitor stuff to game.log, including the original "Checking if we can turn someone into a traitor".

Comment (Jared-Fogle) [reply code/game/gamemodes/dynamic/dynamic_rulesets_midround.dm:183]: Are you sure it's antags, though?

`var/antag_count = mode.current_players[CURRENT_LIVING_ANTAGS].len`

Comment (kriskog) [reply code/game/gamemodes/dynamic/dynamic_rulesets_midround.dm]: I guess antags is correct, yeah

**Review by kriskog — APPROVED** [review]
Go for testmerge and number tweaking when you're ready :+1:

**Comment by RaveRadbury** [comment]
Comment for future dynamic: It would be nice if the `check antagonist` panel could have some info and/or controls for dynamic.

**Comment by kriskog** [comment]
> 
> 
> Comment for future dynamic: It would be nice if the `check antagonist` panel could have some info and/or controls for dynamic.

It has a link to the game mode panel for dynamic, hit that if you want dynamic controls and info.

**Comment by optimumtact** [comment]
We need to merge those kinds of panels into one uber round panel

**Comment by Jared-Fogle** [comment]
Tests are currently ongoing as per the playtest schedule. Marking as a draft PR, though in the end I suspect the changes will be number based.

**Comment by Jared-Fogle** [comment]
Playtesting has entered stage 3, and is now TM'd on Sybil and Terry, alongside Bagil.

**Comment by Jared-Fogle** [comment]
Testing will start soon on Manuel (stage 4).

**Comment by Jared-Fogle** [comment]
Not merge ready, but Rohesie yelled at me :(

**Comment by TheBonded** [comment]
is it possible to have a huge onslaught of antags round start and then another huge amount of midround antags if both role high?

**Comment by notfrying1pans** [comment]
Does not work on manuel, during the testmerge round we had: 1 traitor, a revolution, a revenant, a space dragon, multiple changelings, abductors, and heretics all within a 1 1/2 hour round. Way too chaotic for MRP.

**Comment by Jared-Fogle** [comment]
@TheBonded Yes, though it requires a high round threat level. For example, a 100 threat round could have a 50/50 split of roundstart and midrounds, where roundstarts are now balanced around ~25 threat.

@notfrying1pans I'm definitely expecting to make adjustments for Manuel. Can you provide the round ID of that if possible? I think the most likely change for Manuel is either messing with the timers for midrounds, or lowering the chance of injection.

**Comment by notfrying1pans** [comment]
@Jared-Fogle 1553515

**Comment by Jared-Fogle** [comment]
Ok, so here's the details on that round--https://tgstation13.org/parsed-logs/manuel/data/logs/2021/01/31/round-155351/dynamic.json

- The revenant is not from dynamic. It was a secret random event. There's some talk about letting Dynamic veto random events when the round is deemed chaotic enough, but I consider that out of scope for this PR.
- The dragon was 35 minutes in, the previous max. Recently on Manuel I adjusted the midround timer to 25 to 40 minutes (previously 15 to 35).
- The threat level was 67.4, which is on the off end of the bell curve for deciding threat levels--chaos is to be expected (though I can agree that the level that was seen can be problematic for Manuel).
- The midround budget was 43 out of 67.4. This is extremely unusual, especially considering the budget split is weighted towards roundstarts. That's not to say it's unintentional, but it's to say the midround chaos is, again, expected.

Some adjustments can be made if rounds like this occur too frequently (the midround timer change has already been made, and I await feedback on that as time goes on), but ultimately this seems like a series of some very unlikely events.

**Comment by gryffinp** [comment]
My subjective experience with this testmerge has been that midround antags were rare for good reason, and having the shift get crashed and the station rendered uninhabitable at 40 minutes 7/10 times has been miserable.

**Comment by Jared-Fogle** [comment]
Which server do you play on? Do you have specific problematic round IDs?

**Comment by gryffinp** [comment]
I've been playing on Sybil the last few days. I don't really have a smoking gun round id. I'm not seeing rounds that you'd label as actual _failures_ of the system, I can only report a subjective dissatisfaction with the experience.

**Comment by Jared-Fogle** [comment]
This is ready, but I'm waiting on people with the power to do so to override dynamic.json with the contents I've created for dynamic2021.json. I am currently unable to do that.

**Comment by Jared-Fogle** [comment]
Ready.

**Comment by kriskog** [comment]
Lets goooo

**Pull request merged**
